### PR TITLE
Reject Rules for ExternalIP and svc port if no ep

### LINKED ISSUE #44516: Need ICMP reject rules for externalIPs and port as well
**Is this a request for help?**
No

**What keywords did you search in Kubernetes issues before filing this one?** 
externalIPs kube-proxy connection accepted no pods running. icmp reject rules.

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT

**Kubernetes version** (use `kubectl version`):
kubectl version
```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1", GitCommit:"82450d03cb057bab0950214ef122b67c83fb11df", GitTreeState:"clean", BuildDate:"2016-12-14T00:57:05Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.0", GitCommit:"58b7c16a52c03e4a849874602be42ee71afdcab1", GitTreeState:"clean", BuildDate:"2016-12-12T23:31:15Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```
**Environment**:
- **Cloud provider or hardware configuration**:
Bare Metal.
- **OS** (e.g. from /etc/os-release):
```
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```
- **Kernel** (e.g. `uname -a`):
```
Linux master-1 4.4.0-64-generic #85-Ubuntu SMP Mon Feb 20 11:50:30 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

- **Install tools**:
- **Others**:

**What happened**:
1. Service is exposed using externalIP construct. 
2. No pods are backing the service (using scale deployment 0). The pod is expected to run a TCP based server (non-http)
3. When client connects to this service using externalIP as dest-ip; connection gets accepted by kube-proxy.

**What you expected to happen**:
When no pods are backing service; the connection should have been rejected and definitely not accepted by kube-proxy.

**How to reproduce it** (as minimally and precisely as possible):
1. Create a service and expose it using externalIP construct
```
apiVersion: v1
kind: Service
metadata:
  labels:
    serviceName: my_svc
  name: my_svc
  namespace: xyz
spec:
  clusterIP: 10.100.200.6
  deprecatedPublicIPs:
  - 10.200.0.160
  externalIPs:
  - 10.200.0.160
  ports:
  - name: my_svc_port
    port: 4342
    protocol: TCP
    targetPort: 4342
  - name: my_svc_udp_port
    port: 4342
    protocol: UDP
    targetPort: 4342
  selector:
    serviceName: my_svc
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

2. If pods are present; bring them down using"kubectl scale deployment my_svc --replicas 0 --namespace xyz

3. Client initiates the connection to this service using externalIP and connection goes fine but to kube-proxy.

**Anything else we need to know**:
Info on iptable rules -

When service and pods are present the rules like these -
```
# iptables-save | grep my_svc
  1 -A KUBE-SEP-GQPOLT6NKZE7NNSU -s 10.168.0.44/32 -m comment --comment "xyz/my_svc:my_svc_udp_port" -j KUBE-MARK-MASQ
  2 -A KUBE-SEP-GQPOLT6NKZE7NNSU -p udp -m comment --comment "xyz/my_svc:my_svc_udp_port" -m udp -j DNAT --to-destination 10.168.0.44:4342                                       3 -A KUBE-SEP-LLTPTLEOEOV7X5WW -s 10.168.0.44/32 -m comment --comment "xyz/my_svc:my_svc_port" -j KUBE-MARK-MASQ
  4 -A KUBE-SEP-LLTPTLEOEOV7X5WW -p tcp -m comment --comment "xyz/my_svc:my_svc_port" -m tcp -j DNAT --to-destination 10.168.0.44:4342                                           5 -A KUBE-SERVICES -d 10.100.200.6/32 -p udp -m comment --comment "xyz/my_svc:my_svc_udp_port cluster IP" -m udp --dport 4342 -j KUBE-SVC-FKYB3VE56PGE3NL6
  6 -A KUBE-SERVICES -d 10.200.0.246/32 -p udp -m comment --comment "xyz/my_svc:my_svc_udp_port external IP" -m udp --dport 4342 -j KUBE-MARK-MASQ                               7 -A KUBE-SERVICES -d 10.200.0.246/32 -p udp -m comment --comment "xyz/my_svc:my_svc_udp_port external IP" -m udp --dport 4342 -m physdev ! --physdev-is-in -m addrtype ! --     src-type LOCAL -j KUBE-SVC-FKYB3VE56PGE3NL6                                                                                                                                  8 -A KUBE-SERVICES -d 10.200.0.246/32 -p udp -m comment --comment "xyz/my_svc:my_svc_udp_port external IP" -m udp --dport 4342 -m addrtype --dst-type LOCAL -j KUBE-SVC-         FKYB3VE56PGE3NL6                                                                                                                                                             9 -A KUBE-SERVICES -d 10.100.200.6/32 -p tcp -m comment --comment "xyz/my_svc:my_svc_port cluster IP" -m tcp --dport 4342 -j KUBE-SVC-ACSYPHD3G44RMCBY
 10 -A KUBE-SERVICES -d 10.200.0.246/32 -p tcp -m comment --comment "xyz/my_svc:my_svc_port external IP" -m tcp --dport 4342 -j KUBE-MARK-MASQ                                  11 -A KUBE-SERVICES -d 10.200.0.246/32 -p tcp -m comment --comment "xyz/my_svc:my_svc_port external IP" -m tcp --dport 4342 -m physdev ! --physdev-is-in -m addrtype ! --src-     type LOCAL -j KUBE-SVC-ACSYPHD3G44RMCBY                                                                                                                                     12 -A KUBE-SERVICES -d 10.200.0.246/32 -p tcp -m comment --comment "xyz/my_svc:my_svc_port external IP" -m tcp --dport 4342 -m addrtype --dst-type LOCAL -j KUBE-SVC-             ACSYPHD3G44RMCBY                                                                                                                                                            13 -A KUBE-SVC-ACSYPHD3G44RMCBY -m comment --comment "xyz/my_svc:my_svc_port" -j KUBE-SEP-LLTPTLEOEOV7X5WW
 14 -A KUBE-SVC-FKYB3VE56PGE3NL6 -m comment --comment "xyz/my_svc:my_svc_udp_port" -j KUBE-SEP-GQPOLT6NKZE7NNSU
```

2. When pods are not present; iptables rules look like this -
```
# iptables-save | grep my_svc
  2 -A KUBE-SERVICES -d 10.100.200.6/32 -p udp -m comment --comment "xyz/my_svc:my_svc_udp_port cluster IP" -m udp --dport 4342 -j KUBE-SVC-FKYB3VE56PGE3NL6
  3 -A KUBE-SERVICES -d 10.200.0.246/32 -p udp -m comment --comment "xyz/my_svc:my_svc_udp_port external IP" -m udp --dport 4342 -j KUBE-MARK-MASQ
  4 -A KUBE-SERVICES -d 10.200.0.246/32 -p udp -m comment --comment "xyz/my_svc:my_svc_udp_port external IP" -m udp --dport 4342 -m physdev ! --physdev-is-in -m addrtype ! --src-type LOCAL -j KUBE-      SVC-FKYB3VE56PGE3NL6                                                                                                                                                                                 5 -A KUBE-SERVICES -d 10.200.0.246/32 -p udp -m comment --comment "xyz/my_svc:my_svc_udp_port external IP" -m udp --dport 4342 -m addrtype --dst-type LOCAL -j KUBE-SVC-FKYB3VE56PGE3NL6
  6 -A KUBE-SERVICES -d 10.100.200.6/32 -p tcp -m comment --comment "xyz/my_svc:my_svc_port cluster IP" -m tcp --dport 4342 -j KUBE-SVC-ACSYPHD3G44RMCBY                                                 7 -A KUBE-SERVICES -d 10.200.0.246/32 -p tcp -m comment --comment "xyz/my_svc:my_svc_port external IP" -m tcp --dport 4342 -j KUBE-MARK-MASQ
  8 -A KUBE-SERVICES -d 10.200.0.246/32 -p tcp -m comment --comment "xyz/my_svc:my_svc_port external IP" -m tcp --dport 4342 -m physdev ! --physdev-is-in -m addrtype ! --src-type LOCAL -j KUBE-SVC-      ACSYPHD3G44RMCBY
  9 -A KUBE-SERVICES -d 10.200.0.246/32 -p tcp -m comment --comment "xyz/my_svc:my_svc_port external IP" -m tcp --dport 4342 -m addrtype --dst-type LOCAL -j KUBE-SVC-ACSYPHD3G44RMCBY                  10 -A KUBE-SERVICES -d 10.100.200.6/32 -p udp -m comment --comment "xyz/my_svc:my_svc_udp_port has no endpoints" -m udp --dport 4342 -j REJECT --reject-with icmp-port-unreachable
 11 -A KUBE-SERVICES -d 10.100.200.6/32 -p tcp -m comment --comment "xyz/my_svc:my_svc_port has no endpoints" -m tcp --dport 4342 -j REJECT --reject-with icmp-port-unreachable
```

3. Note that the reject rule exist only for the destination=ServiceIP=10.100.200.6/32

4. When no pods are present; client gets connected to kube-proxy
On VM -
```
~# netstat -napt | grep 4342 | grep 10.250.255.1
tcp        0      0 10.200.0.246:4342       10.250.255.1:54764      ESTABLISHED -

:~# netstat -napt | grep 4342 | grep LISTEN
tcp        0      0 10.200.0.246:4342       0.0.0.0:*               LISTEN      8080/hyperkube
```

---

Based on #43415 and #43969; I manually installed reject rule on INPUT chain. I do not see the issue happening again - meaning client then cannot connect to the server when no pods are present. 

```
-A INPUT -d 10.200.0.246/32 -p tcp -m comment --comment "xyz/my_svc:my_svc_port external IP no endpoint" -m tcp --dport 4342 -j REJECT --reject-with icmp-port-unreachable
```

### PR DESCRIPTION
**Pull request by ketkulka**
- Install ICMP Reject Rules for externalIP and svc port
  if no endpoints are present
- Includes Unit Test case
- Fixes #44516 

**What this PR does / why we need it**:
Explained in issue #44516 
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
`Fixes #44516`

**Special notes for your reviewer**:
Similar to #43415 
Feedback welcome. Will be happy to improve the patch. 
Unit Test done and passing. 

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by thockin** [comment]
@k8s-bot ok to test

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -1017,6 +1017,20 @@ func (proxier *Proxier) syncProxyRules(reason syncReason) {
 			// Allow traffic bound for external IPs that happen to be recognized as local IPs to stay local.
 			// This covers cases like GCE load-balancers which get added to the local routing table.
 			writeLine(natRules, append(dstLocalOnlyArgs, "-j", string(svcChain))...)
+
+			// If the service has no endpoints then reject packets coming from externalIP
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: s/from/via

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -1017,6 +1017,20 @@ func (proxier *Proxier) syncProxyRules(reason syncReason) {
...
+			// Install ICMP Reject rule in filter table for destination=externalIP and dport=svcport
+			if len(newEndpoints[svcName]) == 0 {
+				writeLine(filterRules,
+					"-A", string(kubeServicesChain),
+					"-m", "comment", "--comment", fmt.Sprintf(`"%s has no endpoints from externalIPs"`, svcNameString),
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: s/from externalIPs// - simpler is better

File: pkg/proxy/iptables/proxier.go [hunk 3]
```diff
@@ -1017,6 +1017,20 @@ func (proxier *Proxier) syncProxyRules(reason syncReason) {
...
+					"-m", "addrtype", "--dst-type", "LOCAL",
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: why are you checking addrtype?

**Review by thockin — COMMENTED** [review]
We talked about adding a utility KUBE-MARK-REJECT rule (like KUBE-MARK-DROP) and simply inserting that into the service-chain, which would simplify and catch all these cases.

@freehan did we kill that idea?

**Comment by cblecker** [comment]
@ketkulka:
```
I0417 12:41:49.583] Verifying hack/make-rules/../../hack/verify-gofmt.sh
I0417 12:42:02.606] diff ./pkg/proxy/iptables/proxier_test.go gofmt/./pkg/proxy/iptables/proxier_test.go
I0417 12:42:02.606] --- /tmp/gofmt965435796	2017-04-17 12:41:50.807482217 -0700
I0417 12:42:02.606] +++ /tmp/gofmt953671651	2017-04-17 12:41:50.807482217 -0700
I0417 12:42:02.606] @@ -775,9 +775,9 @@
I0417 12:42:02.607]  			svc.Spec.ClusterIP = svcIP
I0417 12:42:02.607]  			svc.Spec.ExternalIPs = []string{svcExternalIPs}
I0417 12:42:02.607]  			svc.Spec.Ports = []api.ServicePort{{
I0417 12:42:02.607] -				Name: svcPortName.Port,
I0417 12:42:02.607] -				Port: int32(svcPort),
I0417 12:42:02.607] -				Protocol: api.ProtocolTCP,
I0417 12:42:02.607] +				Name:       svcPortName.Port,
I0417 12:42:02.607] +				Port:       int32(svcPort),
I0417 12:42:02.607] +				Protocol:   api.ProtocolTCP,
I0417 12:42:02.607]  				TargetPort: intstr.FromInt(svcPort),
I0417 12:42:02.608]  			}}
I0417 12:42:02.608]  		}),
I0417 12:42:02.608] FAILED   hack/make-rules/../../hack/verify-gofmt.sh	13s
```
Needs a `gofmt -w -s pkg/proxy/iptables/proxier_test.go` run.

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (ketkulka) [inline pkg/proxy/iptables/proxier.go]: fixed in new version

File: pkg/proxy/iptables/proxier.go (hunk 2 above)
Comment (ketkulka) [inline pkg/proxy/iptables/proxier.go]: fixed in new version

File: pkg/proxy/iptables/proxier.go (hunk 3 above)
Comment (ketkulka) [inline pkg/proxy/iptables/proxier.go]: removed addrtype. fixed in new version

**Comment by thockin** [comment]
/approve

@freehan or @dcbw for final

**Comment by ketkulka** [comment]
ping @freehan @dcbw

**Comment by freehan** [comment]
Thanks! /lgtm

**Comment by ketkulka** [comment]
Thanks for approving.
I think the label was not added properly.

Thanks
Ketan

**Comment by cblecker** [comment]
/lgtm
Adding on @freehan's behalf :)

**Comment by freehan** [comment]
@k8s-bot test this

**Comment by cblecker** [comment]
```
I0421 17:41:49.199] k8s.io/kubernetes/pkg/proxy/iptables/proxier.go:1105: undefined: newEndpoints
```

**Comment by ketkulka** [comment]
got conflict with 2f250435fd34151bf4c00fab86f9ea3ff29c12d1 
will submit a corrected patch.

**Comment by ketkulka** [comment]
uploaded a new patch. hopefully this passes.

**Comment by ketkulka** [comment]
Thanks!

**Comment by cblecker** [comment]
@ketkulka I'd make sure you've rebased off master, and that `go test ./pkg/proxy/iptables` works locally.

**Comment by ketkulka** [comment]
actually i think that's what i did and it passed locally. 

git remote add upstream https://github.com/kubernetes/kubernetes.git
git fetch upstream
git rebase upstream/master

i will do it once again. 
let me know if i am mistaken

**Comment by freehan** [comment]
@k8s-bot cvm gce e2e test this
@k8s-bot unit test this
@k8s-bot bazel test this

**Comment by ketkulka** [comment]
Looking into failures. Will send a new patch. 
Sorry for the trouble. Looks like I missed one more line in tests.

**Comment by ketkulka** [comment]
uploaded a new patch. 
this time on my local test; i saw this OK for iptable test 

ok  	k8s.io/kubernetes/pkg/proxy/iptables	0.072s

Apologies for too many small fixes.

**Comment by cblecker** [comment]
okay, main tests are passing now, and it looks good locally for me too
/lgtm

**Comment by ketkulka** [comment]
should this be picked for upcoming release?

**Comment by cblecker** [comment]
@ketkulka This should make it into 1.7

**Pull request merged**